oprf/group/group.go

```diff
@@ -0,0 +1,188 @@
+
+	// Performs the addition operation on the calling Point object
+	// along with a separate Point provided as input
+	Add(*Point) (*Point, error)
+
+	// Performs the negation operation on the calling Point object
+	Neg() (*Point, error)
+
+	// Serializes the Point into a byte slice
+	Serialize() ([]byte, error)
+
+	// Attempts to deserialize a byte slice into a Point
+	Deserialize([]byte) error
+
+	// Returns a bool indicating whether two Points are equal
+	Equal(*Point) bool
+}
+
+// NewPoint generates a new point.
+func NewPoint(c *Ciphersuite) *Point {
+	p := &Point{c, new(big.Int), new(big.Int)}
+	return p
+}
+
+// IsValid checks that the given Point is a valid curve point
+func (p *Point) IsValid() bool {
+	return p.c.curve.IsOnCurve(p.x, p.y)
+}
+
+// ScalarMult multiplies a Point by the provided Scalar value
+func (p *Point) ScalarMult(s *Scalar) (*Point, error) {
+	if !p.IsValid() {
+		return nil, errors.New("invalid point")
+	}
+
+	q := NewPoint(p.c)
+	q.x, q.y = p.c.curve.ScalarMult(p.x, p.y, s.Serialize())
+
+	return q, nil
+}
+
+// Add performs the addition operation on the calling Point object
+// along with a separate Point provided as input
+func (p *Point) Add(q *Point) (*Point, error) {
+	if !p.IsValid() || !q.IsValid() {
+		return nil, errors.New("invalid point")
+	}
+
+	r := NewPoint(p.c)
+	r.x, r.y = p.c.curve.Add(p.x, p.y, q.x, q.y)
+
+	return r, nil
+}
+
+// Neg performs the negation operation on the calling Point object
+// TODO: check opaque to see what is needed
+func (p *Point) Neg() (*Point, error) {
+	if !p.IsValid() {
+		return nil, errors.New("invalid point")
+	}
+
+	r := NewPoint(p.c)
+	return r, nil
+}
+
+// Serialize the Point into a byte slice.
+// TODO: does not handle compressed points.. do we need it?
+func (p *Point) Serialize() ([]byte, error) {
+	if !p.IsValid() {
+		return nil, errors.New("invalid point")
+	}
+
+	x, y := p.x.Bytes(), p.y.Bytes()
+
+	// append zeroes to the front if the bytes are not filled up
+	x = append(make([]byte, p.c.ByteLength()-len(x)), x...)
+	y = append(make([]byte, p.c.ByteLength()-len(y)), y...)
+
+	b := append(x, y...)
+	tag := 4
+
+	return append([]byte{byte(tag)}, b...), nil
+}
+
+// Deserialize an octet-string into a valid Point object.
+func (p *Point) Deserialize(in []byte) error {
+	byteLength := p.c.ByteLength()
+
+	p.x = new(big.Int).SetBytes(in[1 : byteLength+1])
+	p.y = new(big.Int).SetBytes(in[byteLength+1:])
+
+	if !p.IsValid() {
+		return errors.New("invalid point")
+	}
+
+	return nil
+}
+
+// Equal returns a bool indicating whether two Points are equal.
+func (p *Point) Equal(q *Point) bool {
+	if !p.IsValid() || !q.IsValid() {
+		return false
+	}
+
+	return (p.x.Cmp(q.x) == 0) && (p.y.Cmp(q.y) == 0)
+}
+
+// Scalar is an struct representing a field element
+type Scalar struct {
+	c *Ciphersuite
+	x *big.Int
+}
+
+// FieldElement is the interface that represents scalars in a given
+// Ciphersuite instantiation.
+type FieldElement interface {
```

Comment:
This name is confusing, I suggest using Scalar for the interface and `scalar`  for the structure.

Reply:
Changed ;)